landlab/components/soil_grading/soil_grading.py

```diff
@@ -0,0 +1,540 @@
+        """The name of fragmentation pattern"""
+        return self._grading_name
+
+    def create_transition_mat(self):
+        """
+        This procedure creates a transition matrix that control the
+        weathering / fragmentation pattern. The matrix represent the
+        proportion of mass that weathered from each size class to other classes
+        """
+
+        self._A = np.zeros((self._n_sizes, self._n_sizes))
+        precents = np.array(
+            [
+                float(s)
+                for s in self._grading_name.split("-")
+                if s.replace(".", "", 1).isdigit()
+            ]
+        )
+        if "spread" in self._grading_name:
+            precents_to_add = (
+                np.ones((1, self._n_sizes)) * self._precent_of_volume_in_spread
+            )
+            precents = np.append(precents, precents_to_add)
+        alphas_fractios = precents / 100
+        self._A_factor = np.ones_like(self._A) * self._A_factor
+
+        for i in range(self._n_sizes):
+
+            if i == 0:
+                self._A[i, i] = 0
+            elif i == self._n_sizes:
+                self._A[i, i] = -(self._alpha - (self._alpha * alphas_fractios[0]))
+            else:
+                self._A[i, i] = -(self._alpha - (self._alpha * alphas_fractios[0]))
+                cnti = i - 1  # rows,
+                cnt = 1
+                while cnti >= 0 and cnt <= (len(alphas_fractios) - 1):
+                    self._A[cnti, i] = self._alpha * alphas_fractios[cnt]
+                    if cnti == 0 and cnt <= (len(alphas_fractios) - 1):
+                        self._A[cnti, i] = (1 - alphas_fractios[0]) - np.sum(
+                            alphas_fractios[1:cnt]
+                        )
+                        cnt += 1
+                        cnti -= 1
+                    cnt += 1
+                    cnti -= 1
+
+    def set_grading_classes(self):
+
+        input_sizes_flag = self._input_sizes_flag
+
+        def lower_limit_of(maxsize):
+            lower_limit = maxsize * (1 / self._N) ** (power_of)
+            return lower_limit
+
+        upperlimits = []
+        lowerlimits = []
+        num_of_size_classes_plusone = self._n_sizes + 1
+
+        if input_sizes_flag:
+            meansizes = self._meansizes
+            self._meansizes = np.array(meansizes)
+            self._upperlims = np.array(meansizes)
+            self._lowerlims = np.insert(np.array(meansizes), 0, 0)
+        else:
+            maxsize = self._grain_max_size
+            power_of = self._power_of
+            for _ in range(num_of_size_classes_plusone - 1):
+                upperlimits.append(maxsize)
+                maxsize = lower_limit_of(maxsize)
+                lowerlimits.append(maxsize)
+
+            self._upperlims = np.sort(upperlimits)
+            self._lowerlims = np.sort(lowerlimits)
+            self._meansizes = (
+                np.array(self._upperlims) + np.array(self._lowerlims)
+            ) / 2
```

Comment:
How about just saving a `limits` array rather than both the upper and lower limits?

So rewrite as,
```python
        if self._input_sizes_flag:
            self._limits = np.insert(self._meansizes, 0, 0.0)
        else:
            limits = [self._grain_max_size]
            for _ in range(0, self._n_sizes):
                limits.append(limits[-1] * (1 / self._N) ** (self._power_of))
            self._limits = np.asarray(limits[::-1])
        self._meansizes = (self._limits[1:] + self._limits[:-1]) * 0.5
```

Reply:
Sounds good.